crates/data_components/src/databricks/sql_warehouse.rs

```diff
@@ -0,0 +1,655 @@
+use serde_json::{Value, json};
+use snafu::{Snafu, prelude::*};
+use std::{io::Cursor, pin::Pin, sync::Arc};
+use token_provider::TokenProvider;
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Not implemented"))]
+    NotImplemented,
+
+    #[snafu(display("HTTP client build failed: {source}"))]
+    ClientBuildFailed { source: reqwest::Error },
+
+    #[snafu(display("Databricks datatype {ty} not supported"))]
+    UnsupportedType { ty: String },
+
+    #[snafu(display("Unable to retrieve schema: {reason}"))]
+    UnableToRetrieveSchema { reason: String },
+
+    #[snafu(display("HTTP request failed: {source}"))]
+    HttpRequestFailed { source: reqwest::Error },
+
+    #[snafu(display("JSON parsing failed: {source}"))]
+    JsonParsingFailed { source: reqwest::Error },
+
+    #[snafu(display("Missing JSON field: {field}"))]
+    MissingJsonField { field: String },
+
+    #[snafu(display("Invalid JSON array: {field}"))]
+    InvalidJsonArray { field: String },
+
+    #[snafu(display("Failed to deserialize external link: {source}"))]
+    DeserializeExternalLinkFailed { source: serde_json::Error },
+
+    #[snafu(display("Failed to read Arrow stream: {source}"))]
+    ArrowStreamReadFailed { source: arrow::error::ArrowError },
+
+    #[snafu(display("Failed to create table provider: {source}"))]
+    TableProviderCreationFailed { source: DataFusionError },
+
+    #[snafu(display("Failed to initialize SQL table: {source}"))]
+    SqlTableInitializationFailed {
+        source: datafusion_table_providers::sql::sql_provider_datafusion::Error,
+    },
+
+    #[snafu(display("A fully-qualified path is required: {reason}"))]
+    FullyQualifiedPath { reason: String },
+}
+
+/// Main struct for interacting with Databricks SQL Warehouse
+pub struct DatabricksSqlWarehouse {
+    pool: Arc<dyn DbConnectionPool<Arc<SqlWarehouseApi>, &'static (dyn Sync)> + Send + Sync>,
+}
+
+impl DatabricksSqlWarehouse {
+    /// Creates a new Databricks SQL Warehouse instance
+    pub fn new(
+        endpoint: &str,
+        sql_warehouse_id: &str,
+        token_provider: Arc<dyn TokenProvider>,
+    ) -> Result<Self, Error> {
+        let api = Arc::new(SqlWarehouseApi::new(
+            endpoint,
+            sql_warehouse_id,
+            token_provider,
+        )?);
+        let pool = Arc::new(SqlWarehouseConnectionPool { api });
+        Ok(Self { pool })
+    }
+}
+
+struct SqlWarehouseConnectionPool {
+    api: Arc<SqlWarehouseApi>,
+}
+
+#[async_trait]
+impl DbConnectionPool<Arc<SqlWarehouseApi>, &'static dyn Sync> for SqlWarehouseConnectionPool {
+    async fn connect(
+        &self,
+    ) -> Result<
+        Box<dyn DbConnection<Arc<SqlWarehouseApi>, &'static dyn Sync>>,
+        Box<dyn std::error::Error + Send + Sync>,
+    > {
+        Ok(Box::new(SqlWarehouseConnection {
+            api: Arc::clone(&self.api),
+        }))
+    }
+
+    fn join_push_down(&self) -> JoinPushDown {
+        JoinPushDown::Disallow
```

Comment:
This can come in a follow-up, but I think we should allow join push down here for the same endpoint and warehouse id?